sql/plan/walk.go

```diff
@@ -0,0 +1,108 @@
+package plan
+
+import (
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+	"gopkg.in/src-d/go-mysql-server.v0/sql/expression"
+)
+
+// Visitor visits nodes in the plan.
+type Visitor interface {
+	// Visit method is invoked for each node encountered by Walk.
+	// If the result Visitor is not nul, Walk visits each of the children
+	// of the node with that visitor, followed by a call of Visit(nil)
+	// to the returned visitor.
+	Visit(node sql.Node) Visitor
+}
+
+// Walk traverses the plan tree in depth-first order. It starts by calling
+// v.Visit(node); node must not be nil. If the visitor returned by
+// v.Visit(node) is not nil, Walk is invoked recursively with the returned
+// visitor for each children of the node, followed by a call of v.Visit(nil)
+// to the returned visitor.
+func Walk(v Visitor, node sql.Node) {
+	if v = v.Visit(node); v == nil {
+		return
+	}
+
+	for _, child := range node.Children() {
+		Walk(v, child)
+	}
+
+	v.Visit(nil)
+}
+
+type inspector func(sql.Node) bool
+
+func (f inspector) Visit(node sql.Node) Visitor {
+	if f(node) {
+		return f
+	}
+	return nil
+}
+
+// Inspect traverses the plan in depth-first order: It starts by calling
+// f(node); node must not be nil. If f returns true, Inspect invokes f
+// recursively for each of the children of node, followed by a call of
+// f(nil).
+func Inspect(node sql.Node, f func(sql.Node) bool) {
+	Walk(inspector(f), node)
+}
+
+// WalkExpressions traverses the plan and calls expression.Walk on any
+// expression it finds.
+func WalkExpressions(v expression.Visitor, node sql.Node) {
```

Comment:
The content of this function is a little bit specific, right? Maybe we should add a method on Expression interface to return all the expressions on that Expression implementation.

Reply:
`Expression` already has the `Children` method to do that. The problem is that you can't access the expressions of a `Node` because you'd need to do a lot of work every single time you want to do that (just like this function does).

Reply:
Then we need something to avoid have new expressions in the future that are not implemented here, causing a bad walking result. What do you think?

Reply:
We could have an `Expressioner` interface with the method `Expressions` that does the same as `Children` but returning expressions. Instead of forcing it on the `Node` interface I would make it optional, so only the types that have expressions need to implement it.

Reply:
Sounds good, any thoughts @jfontan @mcarmonaa ?